collectors/timex.plugin/plugin_timex.c

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "plugin_timex.h"
+#include "sys/timex.h"
+
+#define PLUGIN_TIMEX_NAME "timex.plugin"
+
+#define CONFIG_SECTION_TIMEX "plugin:timex"
+
+static void timex_main_cleanup(void *ptr)
+{
+    struct netdata_static_thread *static_thread = (struct netdata_static_thread *)ptr;
+    static_thread->enabled = NETDATA_MAIN_THREAD_EXITING;
+
+    info("cleaning up...");
+
+    static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
+}
+
+void *timex_main(void *ptr)
+{
+    netdata_thread_cleanup_push(timex_main_cleanup, ptr);
+
+    int vdo_cpu_netdata = config_get_boolean(CONFIG_SECTION_TIMEX, "timex plugin resource charts", CONFIG_BOOLEAN_YES);
+
+    int update_every = (int)config_get_number(CONFIG_SECTION_TIMEX, "update every", 10);
+    if (update_every < localhost->rrd_update_every)
+        update_every = localhost->rrd_update_every;
+
+    int do_sync = config_get_boolean(CONFIG_SECTION_TIMEX, "clock synchronization state", CONFIG_BOOLEAN_YES);
+    int do_offset = config_get_boolean(CONFIG_SECTION_TIMEX, "time offset", CONFIG_BOOLEAN_YES);
+
+    if (unlikely(do_sync == CONFIG_BOOLEAN_NO && do_offset == CONFIG_BOOLEAN_NO))
+        return NULL;
+
+    usec_t step = update_every * USEC_PER_SEC;
+    heartbeat_t hb;
+    heartbeat_init(&hb);
+    while (!netdata_exit) {
+        usec_t duration = heartbeat_monotonic_dt_to_now_usec(&hb);
+        heartbeat_next(&hb, step);
+
+        if (unlikely(netdata_exit))
```

Comment:
Why do we need to check `netdata_exit` in the loop's body?

Reply:
It's the method we use to shut down Netdata gracefully.

Reply:
AFAICT, this variable should be checked before/after getting into a piece of code that will take a _long_ time to complete **and** doesn't contain any cancellation points. The specific check is right after `heartbeat_next` which contains such a cancellation point. Similarly, I'd expect most of the function calls in the loop body to (a) contain more cancellation points, and/or (b) finish quickly enough so that the loop's condition check suffices.

Reply:
Well, I don't know why we have this pattern in all plugins. Maybe historical reasons? 🤷‍♂️

Reply:
That's what I believe as well. I asked just to make sure there's no some other weird reason. In this case, I'd suggest that we remove the checks that are in the body loop and keep the main one in the loop's header.

Reply:
Probably we are doing this, because users can specify a longer period to sleep, and when the plugin wakes up, instead to run everything one more time, it checks that the correct is to leave.

Reply:
`netdata_exit` is set from `cancel_main_threads()`. Prior to setting `netdata_exit` it will call `netdata_thread_cancel` (ie. `pthread_cancel`). All sleep functions that we use are cancellation points. Given this, I think it doesn't matter what is the duration of a sleep call.

Reply:
In most cases the thread will not even see it coming :smiley: The pthread_cancel will kill it (in fact it may be the case that even in the cleanup code the netdata_exit may or may not be set yet)

(some of the loops we use do disable the cancelability and in that case the netdata_exit will help)

We need to address this at some point

Reply:
Removed the checks.